### PR TITLE
🚀 Customisable Image Visualizer

### PR DESCRIPTION
**Pull request by samet-akcay**
## 📝 Description

- This PR introduces the `ImageVisualizer` class to Anomalib, providing a flexible and customizable way to visualize anomaly detection results. The `ImageVisualizer` can be used as a callback in the Anomalib Engine. The subsequent PRs will add the visualizer to the base model, so users will be able to modify it from there as well. 
- Addresses #1917 

## Features

- Visualize multiple fields from `ImageItem`
- Fully customizable field selection and layout
- Adjustable visualization parameters (gap, font size, background color)
- Overlay option to any field (ie. heatmap on image, masks on image etc.)

### 1: Engine-specific visualizer
When not set, Anomalib will use the default visualizer settings, which would be equivalent to the following

```python
from anomalib.data import MVTec
from anomalib.engine import Engine
from anomalib.models import Padim

data = MVTec()
model = Padim()
engine = Engine()
engine.train(datamodule=data, model=model)
```
The code above will produce the following output

image

### 2: Default visualizer
```python
from anomalib.data import MVTec
from anomalib.engine import Engine
from anomalib.models import Padim
from anomalib.visualization.image import ImageVisualizer

data = MVTec()
model = Padim()
visualizer = ImageVisualizer()
engine = Engine(callbacks=[visualizer])
engine.train(datamodule=data, model=model)
```
image

### 3: Customized visualizer
It is possible to overlay any custom field on top of each other. This includes multiple fields, which overall makes this design fully customizable.
```python
visualizer = ImageVisualizer(
    fields=["image", "gt_mask", "anomaly_map"], 
    overlay_fields=[("image", ["anomaly_map"])]
)
```
image

### 4: Modifying overlay appearance
```python
visualizer = ImageVisualizer(
    overlay_fields_config={
        "pred_mask": {"alpha": 0.7, "color": (255, 0, 0), "mode": "fill"},
    },
)
```
image

### 5: Advanced configuration combining multiple customizations
```python
fields_config={
    "anomaly_map": {"colormap": True, "normalize": True}, 
    "pred_mask": {"color": (0, 0, 255)}
}
overlay_fields_config={
    "anomaly_map": {"colormap": True, "normalize": True},
    "pred_mask": {"alpha": 0.7, "mode": "contour"},
}
text_config={"font": "times.ttf", "size": 24, "color": "white", "background": (0, 0, 0, 180)}

visualizer = ImageVisualizer(
    fields=["image", "gt_mask", "anomaly_map", "pred_mask"],
    overlay_fields=[("image", ["anomaly_map"]), ("image", ["pred_mask"])],
    field_size=(384, 384),
    fields_config=fields_config,
    overlay_fields_config=overlay_fields_config,
    text_config=text_config,
    output_dir="./custom_visualizations",
)
```
image

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

**Comment by jpcbertoldo** [comment]
Just a personal preference but since this PR is open: I think it helps when the contours of the mask are overlaid like this for exemple

contour

File: src/anomalib/visualization/image/functional.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+        col = idx % nrow
+        grid_image.paste(img, (col * img_width, row * img_height))
+
+    return grid_image
+
+
+def visualize_field(
+    field: str,
+    value: torch.Tensor,
+    *,  # Mark the following arguments as keyword-only
```
Comment (ashwinvaidya17) [inline src/anomalib/visualization/image/functional.py:330]: Where is this used?

File: src/anomalib/visualization/image/functional.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+    if field == "anomaly_map":
+        image = visualize_anomaly_map(value, normalize=normalize, colormap=colormap)
+
+    return image
+
+
+def visualize_image_item(
+    item: ImageItem,
+    fields: list[str],
+    *,  # Mark the following arguments as keyword-only
```
Comment (ashwinvaidya17) [inline src/anomalib/visualization/image/functional.py:351]: Same

File: src/anomalib/visualization/image/functional.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+                title = convert_to_title_case(field)
+                image = add_text_to_image(image, title)
+                images.append(image)
+
+    # Process overlay fields
+    if overlay_fields:
+        for base, overlays in overlay_fields:
+            if base in field_images:
+                base_image = field_images[base].copy()
+                valid_overlays = [overlay for overlay in overlays if overlay in field_images]
```
Comment (ashwinvaidya17) [inline src/anomalib/visualization/image/functional.py:393]: Should we warn if there are a few invalid overlays?

**Review by ashwinvaidya17 — COMMENTED** [review]
The new visualizer looks amazing. I have a few minor comments. Also, how do we plan to tackle text-only output of VLM?

**Comment by samet-akcay** [comment]
> The new visualizer looks amazing. I have a few minor comments. Also, how do we plan to tackle text-only output of VLM?

This is for image-based models. I'll need to implement `VideoVisualizer`, and possibly something like `VLMVisulizer`

File: src/anomalib/visualization/image/functional.py (hunk 1 above)
Comment (samet-akcay) [reply src/anomalib/visualization/image/functional.py:330]: I want user to use keywords for the rest of the arguments after *. For example, following line follows this keyword-only format 
https://github.com/samet-akcay/anomalib/blob/c17c77ce655cc146e065699bd10f69a9cb9b3ac1/src/anomalib/visualization/image/functional.py#L376

Comment (samet-akcay) [reply src/anomalib/visualization/image/functional.py:330]:
```python
image = visualize_field(field, value, colormap=colormap, normalize=normalize)
```

File: src/anomalib/visualization/image/functional.py (hunk 2 above)
Comment (samet-akcay) [reply src/anomalib/visualization/image/functional.py:351]:
```python
image = visualize_image_item(
                item,
                fields=self.fields,
                overlay_fields=self.overlay_fields,
                alpha=self.alpha,
                colormap=self.colormap,
                normalize=self.normalize,
            )
```

File: src/anomalib/visualization/image/functional.py (hunk 1 above)
Comment (ashwinvaidya17) [reply src/anomalib/visualization/image/functional.py:330]: So, basically we want to force the users to pass parameters using keywords

**Review by ashwinvaidya17 — APPROVED** [review]

**Comment by samet-akcay** [comment]
@jpcbertoldo, how does it look like now? Both is possible with the updated PR

**Review by ashwinvaidya17 — COMMENTED** [review]
Looks fine to me. But not sure if I am completely sold on the field config, and overlay field config part. I usually rely on my IDE to suggest customisable parameters. With the nested dict, it is harder to figure how, and to what I can I change the parameters to.
How often would we see users change these parameters? If it is not very frequent, then I am in favour of being a little more opinionated and keep these parameters as constants. If the users want something else, they can modify the source.

**Comment by samet-akcay** [comment]
> Looks fine to me. But not sure if I am completely sold on the field config, and overlay field config part. I usually rely on my IDE to suggest customisable parameters. With the nested dict, it is harder to figure how, and to what I can I change the parameters to. How often would we see users change these parameters? If it is not very frequent, then I am in favour of being a little more opinionated and keep these parameters as constants. If the users want something else, they can modify the source.

that'd make my life easier :)

**Comment by jpcbertoldo** [comment]
> @jpcbertoldo, how does it look like now? Both is possible with the updated PR

Looks really cool. The design is somehow complete but not complicated as expected haha. 

I wanted to try something like this

```python
visualizer = ImageVisualizer(
    fields=["image",  "anomaly_map"],
    overlay_fields=[("image", ["anomaly_map", "gt_mask"]), ("anomaly_map", ["gt_mask"])],
    fields_config={"pred_mask": {"color": (255, 0, 0)}},
    overlay_fields_config={
        "anomaly_map": {"alpha": 0.6, "mode": "fill"},
        "pred_mask": {"alpha": 1., "mode": "contour"},
    },
)
```

Will try tomorrow, my environment has some unsaved stuff right now. 

By the way the output of example 5 is outdated, right? Because the pred mask config is set to `contour` but we see a `fill` in the image.

File: src/anomalib/visualization/image/functional.py (hunk 1 above)
Comment (samet-akcay) [reply src/anomalib/visualization/image/functional.py:472]: Yes, because such args are hard to read/follow if a keyword is not provided

**Comment by samet-akcay** [comment]
> > @jpcbertoldo, how does it look like now? Both is possible with the updated PR
> 
> Looks really cool. The design is somehow complete but not complicated as expected haha.
> 
> I wanted to try something like this
> 
> ```python
> visualizer = ImageVisualizer(
>     fields=["image",  "anomaly_map"],
>     overlay_fields=[("image", ["anomaly_map", "gt_mask"]), ("anomaly_map", ["gt_mask"])],
>     fields_config={"pred_mask": {"color": (255, 0, 0)}},
>     overlay_fields_config={
>         "anomaly_map": {"alpha": 0.6, "mode": "fill"},
>         "pred_mask": {"alpha": 1., "mode": "contour"},
>     },
> )
> ```
> 
> Will try tomorrow, my environment has some unsaved stuff right now.
> 
> By the way the output of example 5 is outdated, right? Because the pred mask config is set to `contour` but we see a `fill` in the image.

For some reason, `fields_config` is overwriting the `pred_mask` configs in `overlay_fields_config`. I'll investigate that

**Comment by jpcbertoldo** [comment]
Really nice feature!

```python
from anomalib.visualization.image import ImageVisualizer
visualizer = ImageVisualizer(
    overlay_fields=[("image", ["gt_mask"]), ("image", ["anomaly_map", "gt_mask"]), ("anomaly_map", ["gt_mask"])],
    overlay_fields_config={
        "anomaly_map": {"alpha": 0.5, "mode": "fill"},
        "gt_mask": {"alpha": 1., "mode": "contour", "color": (255, 0, 255)},
    },
)
engine = Engine(callbacks=[visualizer])
engine.test(datamodule=data, model=model)
```

gives

image

i would suggest gt_mask overlayed as a default, it really helps to compare the images because it gives a reference shape to locate things

**Comment by SweetWind1996** [comment]
I used the above 5 visualization methods. No matter which one I replaced, the result was consistent with the default effect. Moreover, the anomaly graph is still green in efficientAD. The anomalib version I used is v2. I just modified the example code:
``` python
from anomalib.data import MVTecAD
from anomalib.engine import Engine
from anomalib.models import EfficientAd
from anomalib.visualization import ImageVisualizer

# 1. Basic Usage
# Initialize with default settings
fields_config = {
    "anomaly_map": {"colormap": True, "normalize": True},
    "pred_mask": {"color": (0, 0, 255)},
}
overlay_fields_config = {
    "anomaly_map": {"colormap": True, "normalize": True},
    "pred_mask": {"alpha": 0.7, "mode": "contour"},
}
text_config = {
    "font": "times.ttf",
    "size": 24,
    "color": "white",
    "background": (0, 0, 0, 180),
}

visualizer = ImageVisualizer(
    fields=["image", "gt_mask", "anomaly_map", "pred_mask"],
    overlay_fields=[("image", ["anomaly_map"]), ("image", ["pred_mask"])],
    field_size=(384, 384),
    fields_config=fields_config,
    overlay_fields_config=overlay_fields_config,
    text_config=text_config,
    output_dir="./custom_visualizations",
)

# 2. Custom Configuration
# Configure model parameters

model = EfficientAd(
    teacher_out_channels=384,  # Number of teacher output channels
    model_size="medium",
    lr=1e-4,
)

# 3. Training Pipeline
# Set up the complete training pipeline
datamodule = MVTecAD(
    root="./datasets/MVTecAD", category="bottle", train_batch_size=1, eval_batch_size=1
)

# Initialize training engine with specific settings
engine = Engine(
    max_epochs=200,
    accelerator="auto",  # Automatically detect GPU/CPU
    devices=1,  # Number of devices to use
    callbacks=[visualizer],
)

engine.test(
    model=model,
    datamodule=datamodule,
    ckpt_path="model.ckpt",
)
```

Also, has the visualization method in the official documentation been deprecated？：
https://anomalib.readthedocs.io/en/latest/markdown/guides/how_to/visualization/visualize_image.html

image

**Comment by samet-akcay** [comment]
@SweetWind1996, can you pass your `visualizer` to the model like the following:

```python
model = EfficientAd(
    teacher_out_channels=384,  # Number of teacher output channels
    model_size="medium",
    lr=1e-4,
    visualizer=visualizer,
)
```

**Comment by SweetWind1996** [comment]
@samet-akcay Thanks for your reply! The number of images in the visualization now matches the visualizer's count. However, the anomaly map is still all green. The prediction mask (pred_mask) seems to be fine, so I'm not sure how to fix this. Any ideas?

To add a bit more context, I saw a potential solution in a previous issue that suggested setting normalization=None. The problem is, it seems the config has been removed in v2, so I'm a bit lost on how to apply this setting or what the new approach should be.
Any ideas on how to solve this in v2?

image

**Pull request merged**